include/oneapi/dpl/internal/distributed_ranges_impl/shp/algorithms/exclusive_scan.hpp

```diff
@@ -0,0 +1,223 @@
+        device_allocator<T> allocator(context(), root);
+        vector<T, device_allocator<T>> partial_sums(std::size_t(zipped_segments.size()), allocator);
+
+        segment_id = 0;
+        for (auto&& segs : zipped_segments)
+        {
+            auto&& [in_segment, out_segment] = segs;
+
+            auto&& q = __detail::queue(ranges::rank(in_segment));
+            auto&& local_policy = __detail::dpl_policy(ranges::rank(in_segment));
+
+            auto dist = rng::distance(in_segment);
+            assert(dist > 0);
+
+            auto first = rng::begin(in_segment);
+            auto last = rng::end(in_segment);
+            auto d_first = rng::begin(out_segment);
+
+            auto init = inits[segment_id];
+
+            auto event = exclusive_scan_async(local_policy, dr::__detail::direct_iterator(first),
+                                              dr::__detail::direct_iterator(last),
+                                              dr::__detail::direct_iterator(d_first), init, binary_op);
+
+            auto dst_iter = ranges::local(partial_sums).data() + segment_id;
+
+            auto src_iter = ranges::local(out_segment).data();
+            rng::advance(src_iter, dist - 1);
+
+            auto e = q.submit([&](auto&& h) {
+                h.depends_on(event);
+                h.single_task([=]() {
+                    rng::range_value_t<O> value = *src_iter;
+                    *dst_iter = value;
+                });
+            });
+
+            events.push_back(e);
+
+            segment_id++;
+        }
+
+        __detail::wait(events);
+        events.clear();
+
+        auto&& local_policy = __detail::dpl_policy(0);
+
+        auto first = ranges::local(partial_sums).data();
+        auto last = first + partial_sums.size();
+
+        inclusive_scan_async(local_policy, first, last, first, binary_op).wait();
+
+        std::size_t idx = 0;
+        for (auto&& segs : zipped_segments)
+        {
+            auto&& [in_segment, out_segment] = segs;
+
+            if (idx > 0)
+            {
+                auto&& q = __detail::queue(ranges::rank(out_segment));
+
+                auto first = rng::begin(out_segment);
+                dr::__detail::direct_iterator d_first(first);
+
+                auto d_sum = ranges::__detail::local(partial_sums).begin() + idx - 1;
+
+                sycl::event e = dr::__detail::parallel_for(q, sycl::range<>(rng::distance(out_segment)), [=](auto idx) {
+                    d_first[idx] = binary_op(d_first[idx], *d_sum);
+                });
+
+                events.push_back(e);
+            }
+            idx++;
+        }
+
+        __detail::wait(events);
+    }
+    else
+    {
+        assert(false);
+    }
+}
+
+// Ranges versions
+
+template <typename ExecutionPolicy, distributed_contiguous_range R, distributed_contiguous_range O, typename T,
+          typename BinaryOp>
+void
+exclusive_scan(ExecutionPolicy&& policy, R&& r, O&& o, T init, BinaryOp&& binary_op)
```

Comment:
Is there a reason for passing `binary_op` as a rvalue reference? I think the existing algorithms pass the binary operator by value.

Reply:
There is templated parameter before "&&", so it is so called [universal reference](https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers) - something that can be lvalue or rvalue depending on calling code. It is a safe way to take user arguments by the library.

Reply:
You are correct that it is a universal reference.  I think Julian's point is that throughout oneDPL,I believe our convention is to accept callable parameters by value (binary ops, comparators, predicates, projection ops, etc.).  Unless there is a specific reason to warrant universal references, I think we should probably follow that convention.  
(This is something that we should probably codify into a style or convention guide for oneDPL)

Reply:
There's no particular reason; we'll switch to passing by value.

Reply:
ok, applied to all operators and all comparators